From f27954841b67654efb0b37a37972e5db4b9a7f4a Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Wed, 29 Oct 2014 16:34:25 -0400
Subject: [fs] ext4: validate external journal superblock checksum

Message-id: <1414600485-1929-36-git-send-email-lczerner@redhat.com>
Patchwork-id: 98573
O-Subject: [RHEL 7.1 PATCH v2 35/55] ext4: validate external journal superblock checksum
Bugzilla: 1156096
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: "Darrick J. Wong" <darrick.wong@oracle.com>

BZ 1156096
https://bugzilla.redhat.com/show_bug.cgi?id=1156096

BREW 8169722
https://brewweb.devel.redhat.com/taskinfo?taskID=8169722

Upstream commit df4763bea5b04d8eed941cfe3df51f22cfe95570

 If the external journal device has metadata_csum enabled, verify
 that the superblock checksum matches the block before we try to
 mount.

 Signed-off-by: Darrick J. Wong <darrick.wong@oracle.com>
 Reviewed-by: Jan Kara <jack@suse.cz>
 Signed-off-by: Theodore Ts'o <tytso@mit.edu>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/super.c b/fs/ext4/super.c
index 161d613..6c08ff8 100644
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -4420,6 +4420,15 @@ static journal_t *ext4_get_dev_journal(struct super_block *sb,
   goto out_bdev;
  }
 
+ if ((le32_to_cpu(es->s_feature_ro_compat) &
+      EXT4_FEATURE_RO_COMPAT_METADATA_CSUM) &&
+     es->s_checksum != ext4_superblock_csum(sb, es)) {
+  ext4_msg(sb, KERN_ERR, "external journal has "
+           "corrupt superblock");
+  brelse(bh);
+  goto out_bdev;
+ }
+
  if (memcmp(EXT4_SB(sb)->s_es->s_journal_uuid, es->s_uuid, 16)) {
   ext4_msg(sb, KERN_ERR, "journal UUID does not match");
   brelse(bh);
-- 
1.7.1